File: sinaleasy-back/src/main/java/com/sinalez/sinaleasy_back/entities/Signal.java
package com.sinalez.sinaleasy_back.entities;

import java.io.Serializable;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import jakarta.persistence.Transient;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;
import lombok.Setter;

@Entity
@Table(name = "TB_Signals")
@Getter
@Setter
public class Signal implements Serializable {
    private static final long serialVersionUID = 1L;
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private UUID signalId;

    @NotBlank private String name;
    @NotBlank private String description;
    @NotBlank private String address;
    @NotNull private Double latitude;
    @NotNull private Double longitude;
    @NotNull private Integer typeOfSignal;
    @NotNull private Integer status;
    private LocalDate date; // Mudar esse nome para algo mais sugestivo
    @Transient private Integer numberOfLikes;
    @Transient private Integer scaleFactor;

    @JsonIgnoreProperties("signals")
    @ManyToOne
    @JoinColumn(name = "city_id", nullable = false)
    private City city;

    @JsonIgnoreProperties("signals")
    @ManyToOne
    @JoinColumn(name = "user_id")
    private User user;

    @JsonBackReference
    @OneToMany(mappedBy = "signal", cascade = CascadeType.ALL, orphanRemoval = true) //pq nao funciona
    private List<Milestone> signalMilestones;

    @JsonIgnoreProperties("signals")
    @OneToOne
    @JoinColumn(name = "grade_id")
    private Grade grade;

    @OneToMany(mappedBy = "signal", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<UserSignal> signalVotes; // lista de votos recebidos pelo sinal


    public void setCity(City cityOfSignal) {
        if (cityOfSignal == null) {
            throw new IllegalArgumentException("Cidade não pode ser nula!");
        }
        this.city = cityOfSignal;
    }

    public void setSignalMilestone(Milestone milestone) {
        if(milestone == null) {
            throw new IllegalArgumentException("Novo status não pode ser nulo!");
        }
        if(this.signalMilestones == null){
            this.signalMilestones = new ArrayList<Milestone>();
        }
        this.signalMilestones.add(milestone);
    }

    public void updateStatus(Integer newStatus) {
        if (this.status != newStatus) {
            Milestone newMilestone = new Milestone();
            newMilestone.setStatus(newStatus);
            newMilestone.setStatusUpdateTime(LocalDate.now());
            newMilestone.setSignal(this);
            this.signalMilestones.add(newMilestone);
            this.status = newStatus;
        }
    }


    public void addGradeIfConcluded(Integer rating, String gradeDescription, LocalDate gradeUpdateTime) {
        if (this.status == 3) {
            Grade newGrade = new Grade();
            newGrade.setRating(rating);
            newGrade.setDescription(description);
            newGrade.setDate(gradeUpdateTime);
            newGrade.setSignal(this);
            this.grade = newGrade;
        }
    }

}

File: sinaleasy-front/src/domain/Signal.ts
import * as L from 'leaflet';
import { Milestone } from './Milestone';
import { Grade } from './Grade';

export class Signal {
    signalId: string | undefined;
    name: string | undefined;
    date: Date | undefined;
    address: string | undefined;
    description: string | undefined;
    typeOfSignal: number | undefined;
    latitude: number | undefined;
    longitude: number | undefined;
    scaleFactor: number | undefined;
    status: number | undefined;
    numberOfLikes: number | undefined;
    cityId: string | undefined;
    liked: boolean | undefined;
    signalMilestones: Array<Milestone> | undefined;
    grade: Grade | undefined;
    
    constructor(name?: string, date?: Date, address?: string, description?: string, typeOfSignal?: number, latitude?: number, longitude?: number, scaleFactor?: number, status?: number, numberOfLikes?: number, cityId?: string, signalId?: string, liked?: boolean, signalMilestones?: Array<Milestone>, grade?: Grade) {
        this.name = name;
        this.date = date;
        this.address = address;
        this.description = description;
        this.typeOfSignal = typeOfSignal;
        this.latitude = latitude;
        this.longitude = longitude;
        this.scaleFactor = scaleFactor;
        this.status = status;
        this.numberOfLikes = numberOfLikes;
        this.cityId = cityId;
        this.signalId = signalId;
        this.liked = liked;
        this.signalMilestones = signalMilestones;
        this.grade = grade;
    }

}
